src/defck.rs

```diff
@@ -0,0 +1,471 @@
+                    let f = app(eq, cons(app(y, nil), cons(app(z, nil), nil)));
+                    check_hyp(self.db, iter, f)?;
+                    iter.next().is_none().then_some(())?;
+                    let f = app(eq, cons(app(x, nil), cons(app(z, nil), nil)));
+                    check(self.db, addr, f)
+                })?;
+            }
+            [Keyword(cmd), rest @ ..] if cmd.as_ref(buf) == b"primitive" => {
+                self.flush_pending_syntax();
+                for label in rest {
+                    let primitive = self.nset.lookup_label(label.value(buf)).unwrap().atom;
+                    // Remove the definition from the pending list
+                    if let Some(pending_index) =
+                        self.pending_primitive.iter().position(|x| x.0 == primitive)
+                    {
+                        self.pending_primitive.swap_remove(pending_index);
+                        self.result.primitives.push(primitive)
+                    } else {
+                        self.result.diagnostics.push((
+                            StatementAddress::new(sref.id, index),
+                            Diagnostic::DefCkMisplacedPrimitive(label.span()),
+                        ));
+                    }
+                }
+            }
+            [Keyword(cmd), justif_label, Keyword(for_), label]
+                if cmd.as_ref(buf) == b"justification" && for_.as_ref(buf) == b"for" =>
+            {
+                let theorem = self
+                    .nset
+                    .lookup_label(justif_label.value(buf))
+                    .unwrap()
+                    .atom;
+                let definition = self.nset.lookup_label(label.value(buf)).unwrap().atom;
+                self.result.justifications.insert(definition, theorem);
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+
+    fn verify_definition_statement(
+        &mut self,
+        syntax_axiom: Label,
+        addr: StatementAddress,
+    ) -> Result<(), Diagnostic> {
+        let stmt = self.db.statement_by_address(addr);
+        let definition = self.nset.lookup_label(stmt.label()).unwrap().atom;
+
+        let syntax_addr = self.nset.lookup_label_by_atom(syntax_axiom).address;
+
+        // push the definition to the validated list early, so that later definitions
+        // aren't as messed up if this check fails
+        self.result.definitions.insert(definition);
+        if let Some(prev) = self.result.def_map.insert(syntax_axiom, definition) {
+            return Err(Diagnostic::DefCkDuplicateDefinition(
+                self.nset.atom_name(syntax_axiom).into(),
+                self.nset.lookup_label_by_atom(prev).address,
+            ));
+        }
+
+        if self.result.justifications.contains_key(&definition) {
+            // Skip definitional check for definitions having a justification.
+
+            // TODO check that the justification matches the definition
```

Comment:
So I've actually checked the only instance of `justification` command in set.mm, `bijust` for `df-bi`, and actually in that case the statements are completely different and I believe they would have to be verified manually.

So I don't think anything else needs to be done in this branch and the comment can be removed.

Reply:
I'm off to bed now but I've pushed some code regarding justification handling. I anticipate that df-bi can be handled, although bound variables are complicated (with or without justification handling).